src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1674,6 +1775,7 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 		UpdateRelationToShardNames((Node *) originalQuery, *relationShardList);
```

Comment:
What do you think of moving this line to `SingleShardSelectTaskList` and `SingleShardModifyTaskList`? We also do this in `MultiShardModifyTaskList` for multi-shard UPDATE/DELETE. It might also make sense to define a little function that first does `UpdateRelationToShardNames` and then `pg_get_query_def`.

Reply:
In that case, we need to change the implementation of INSERT INTO .. SELECT queries and subqueries too.